agent/utils/ephemeral_ports.go

```diff
@@ -67,13 +64,35 @@ func GenerateEphemeralPortNumbers(n int, reserved []uint16) ([]uint16, error) {
 	return result, nil
 }
 
+// safePortTracker tracks the host port last assigned to a container port range and is safe to use concurrently.
+// TODO: implement a port manager that does synchronization and integrates with a configurable option to modify ephemeral range
+type safePortTracker struct {
+	mu                   sync.Mutex
+	lastAssignedHostPort int
+}
+
+// SetLastAssignedHostPort sets the last assigned host port
+func (pt *safePortTracker) SetLastAssignedHostPort(port int) {
+	pt.mu.Lock()
+	defer pt.mu.Unlock()
+
+	pt.lastAssignedHostPort = port
+}
+
+// GetLastAssignedHostPort returns the last assigned host port
+func (pt *safePortTracker) GetLastAssignedHostPort() int {
+	pt.mu.Lock()
+	defer pt.mu.Unlock()
+
+	return pt.lastAssignedHostPort
+}
+
 var dynamicHostPortRange = getDynamicHostPortRange
+var tracker safePortTracker
 
 // GetHostPortRange gets N contiguous host ports from the ephemeral host port range defined on the host.
 func GetHostPortRange(numberOfPorts int, protocol string) (string, error) {
-	portLock.Lock()
-	defer portLock.Unlock()
```

Comment:
Discussed offline. The requirement to synchronize access to this function has not changed so we probably still need this lock.

Reply:
added this back. This lock prevents two concurrent task-level go routines from searching for ports at the same time. The port tracker is also being kept atomic in nature, since its a global variable.

also updating the PR description.